File: Hashing_2_Problem_2_ContiguousArray.java
import java.util.HashMap;
//Time Complexity : O(n)
//Space Complexity : O(n) where n=length of the array
// Run Successfully on LeetCode
// Problems : Understanding the edge case : 0 on -1st index

/*
Here we will be storing the value of aggregate sum in hashmap with its index
If that sum repeated somewhere in the array, the difference of last occurence and
current occurence will be the size of the array we need which is equal number of 0s and 1s
 */


public class Hashing_2_Problem_2_ContiguousArray {
    public static int findMaxLength(int[] nums) {
        if(nums==null || nums.length ==0) return 0;
        int maxCount = 0;
        HashMap<Integer,Integer> hm = new HashMap<>();
        hm.put(0,-1);
        int runningSum = 0;
        for(int i=0; i < nums.length ; i++)
        {
            int toAdd = (nums[i]==0) ? -1 : 1;
            runningSum +=toAdd;
            if(hm.containsKey(runningSum))
            {
                maxCount = Math.max(i - hm.get(runningSum),maxCount);
            }
            else
            {
                hm.put(runningSum,i);
            }
        }
        return maxCount;
    }

    public static void main(String[] args) {
        int[] arr = {0,1,0,1};
        System.out.println(findMaxLength(arr));
    }
}

File: Hashing-2-Problem 2- Contiguous Array.py
"""
S30 FAANMG Problem #18 {Medium}


525. Contiguous Array

Time Complexity : O(N)
 

Space Complexity : O(N)


Did this code successfully run on Leetcode : Yes


A hashmap is created to store the running Sum

The array is iterated till the last element
As we have to return the maximum length of a contiguous subarray with an equal number of 0 and 1..

Here 0 is considered to be -1
    and 1 is considered as 1 
    
there is maximum Number that is storing the current max balanced length 

For each iteration the running sum is calculated by adding with the Value as above for the corresponding index
    If the rSum is in the hashmap, 
     then max of the the current maxn and the index - the previous index of the rSum is calculates
     
     else add the corresesponding index of the rSum to the hashmap
  

    
@name: Rahul Govindkumar_RN27JUL2022
"""


class Solution:
    def findMaxLength(self, nums: List[int]) -> int:
        
        hashmap = {0:-1}
        rSum = 0
        maxN = 0
        for i in range (len(nums)):
            
            if(nums[i]==0):
                rSum -= 1
            else:
                rSum += 1
                
            if(rSum in hashmap):
                maxN = max(maxN, i - hashmap[rSum])
            else:
                hashmap[rSum] = i
        return maxN
                
                
        
        